[AtomicExpand] Add bitcasts when expanding load atomic vector

jofrn opened the pull request:
AtomicExpand fails for aligned `load atomic <n x T>` because it
does not find a compatible library call. This change adds appropriate
bitcasts so that the call can be lowered. It also adds support for
128 bit lowering in tablegen to support SSE/AVX.

jofrn committed:
[X86] Cast atomic vectors in IR to support floats

This commit casts floats to ints in an atomic load during AtomicExpand to support
floating point types. It also is required to support 128 bit vectors in SSE/AVX.

jofrn committed:
[AtomicExpand] Add bitcasts when expanding load atomic vector

AtomicExpand fails for aligned `load atomic <n x T>` because it
does not find a compatible library call. This change adds appropriate
bitcasts so that the call can be lowered. It also adds support for
128 bit lowering in tablegen to support SSE/AVX.

jofrn commented:
* **#148900** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/148900?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/llvm/llvm-project/148900?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#148899** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/148899?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#165818** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/165818?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#148898** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/148898?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#148897** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/148897?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#148896** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/148896?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#148895** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/148895?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#148894** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/148894?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#148893** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/148893?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

arsenm reviewed (commented):
I don't understand the reason to bother with this. I want to fully remove the bitcast handling in AtomicExpand; your other patches already do the correct thing in SelectionDAG

jofrn commented on llvm/lib/Target/X86/X86InstrCompiler.td:
```diff
@@ -1211,6 +1211,11 @@ def : Pat<(v4i32 (scalar_to_vector (i32 (atomic_load_32 addr:$src)))),
+def : Pat<(v2i64 (atomic_load_128_v2i64 addr:$src)),
+           (VMOVAPDrm addr:$src)>; // load atomic <2 x i64>
+def : Pat<(v4i32 (atomic_load_128_v4i32 addr:$src)),
+           (VMOVAPDrm addr:$src)>; // load atomic <4 x i32>
```
It is required for the SSE/AVX extensions, like in `atomic_vec2_ptr_align` with `-mcpu=x86-64-v2`. If the extensions are enabled, then this pattern has to be handled in this test.

arsenm commented on llvm/lib/CodeGen/AtomicExpandPass.cpp:
```diff
@@ -483,7 +483,12 @@ LoadInst *AtomicExpandImpl::convertAtomicLoadToIntegerType(LoadInst *LI) {
+  Value *NewVal =
+      LI->getType()->isPointerTy() ||
+              (LI->getType()->isVectorTy() &&
+               cast<VectorType>(LI->getType())->getElementType()->isPointerTy())
```
```suggestion
      LI->getType()->isPtrOrPtrVectorTy()
```

arsenm commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -1933,6 +1933,20 @@ def atomic_load_64 :
+          (atomic_load node:$ptr)> {
+  let IsAtomic = true;
+  let MemoryVT = v4i32;
+}
```
I don't get it. The core change here only changes the IR pass, in such a way that the later codegen pieces do not. It would make more sense if this was the patch to directly handling the casting in SelectionDAG

jofrn commented:
Moved the .td records required for this PR back to prior PRs https://github.com/llvm/llvm-project/pull/148899 (Cast atomic vec with float -- containing the Pats for X86) and https://github.com/llvm/llvm-project/pull/165818 (Split -- containing the definitions of the PatFrag atomics).

jofrn commented on llvm/test/Transforms/AtomicExpand/X86/expand-atomic-non-integer.ll:
```diff
@@ -189,3 +189,69 @@ define void @pointer_cmpxchg_expand6(ptr addrspace(1) %ptr, ptr addrspace(2) %v)
+; CHECK-NEXT:    [[TMP7:%.*]] = inttoptr <2 x i64> [[TMP6]] to <2 x ptr>
+; CHECK-NEXT:    ret <2 x ptr> [[TMP7]]
+;
+  %ret = load atomic <2 x ptr>, ptr %x acquire, align 16
```
I think we should. Changed it to that; however, some \*\_cmpxchg\_\* tests also have their output changed even though unrelated to expanding a load atomic vector, so perhaps it should be a separate commit. What do you think?

RKSimon replied:
Yes, please pre-commit the additional tests and RUN line so we can see the codegen diff in this PR

jofrn replied:
I have done so already within this PR. Take a look please.

RKSimon reviewed (approved):
LGTM

jofrn committed:
[SelectionDAG] Split vector types for atomic load

Vector types that aren't widened are split
so that a single ATOMIC_LOAD is issued for the entire vector at once.
This change utilizes the load vectorization infrastructure in
SelectionDAG in order to group the vectors. This enables SelectionDAG
to translate vectors with type bfloat,half.

jofrn committed:
Merge branch 'main' into users/jofrn/gt/07-15-_atomicexpand_add_bitcasts_when_expanding_load_atomic_vector
